server/registry/server.go

```diff
@@ -71,7 +71,15 @@ func New(config Config) *RegistryServer {
 		s.dbConfig = "/tmp/registry.db"
 	}
 
-	return s
+	db, err := storage.NewClient(context.Background(), s.database, s.dbConfig)
+	if err != nil {
+		return nil, err
+	}
+	defer db.Close()
+	if err := db.EnsureTables(); err != nil {
+		return nil, err
+	}
```

Comment:
I still think constructors should only construct their return type - setting fields and applying options/defaults. The fact that this constructor returns an error and mutates the database indicates it's doing much more than just constructing. It would be more accurately named `NewAndInitialize` with this addition.

If we're going to bump this into a future PR, let's make sure to actually do that soon. It seems like a really easy change to split this into a separate initialize method for this PR though.

Reply:
I appreciate the aesthetic, and I found [someone who shares it](https://www.reddit.com/r/golang/comments/9bkql2/stop_returning_errors_in_constructors_use/) but I think the followup discussion there suggests that it's not universal or always practical. There are cases (a few, but some) where constructors in the standard library return errors, and I don't (personally) think that creating a server needs to exclude some assurance that the server being returned is valid - that's what's happening [here](https://cs.opensource.google/go/go/+/refs/tags/go1.17.3:src/net/http/request.go;l=863)  and [here](https://cs.opensource.google/go/go/+/refs/tags/go1.17.3:src/archive/zip/reader.go;l=87).